src/server/storage_engine/default_engine.rs

```diff
@@ -180,10 +211,58 @@ impl StorageEngine for DefaultEngine {
             }
         }
         match self.dir_db.db.get(path.as_bytes())? {
-            Some(value) => {
-                // debug!("read_dir path: {}, value: {:?}", path, value);
-                Ok(value)
-            }
+            Some(value) => match bincode::deserialize::<SubDirectory>(&value) {
+                Ok(sub_dir) => {
+                    let mut iter = sub_dir.sub_dir.iter().skip(offset as usize);
+                    let result = vec![0u8; size as usize];
+                    let mut dirp_ptr = result.as_slice().as_ptr();
+                    let mut total = 0;
+                    loop {
+                        match iter.next() {
+                            Some(value) => {
+                                let reclen =
+                                    offset_of!(dirent64, d_name) as u16 + value.0.len() as u16 + 1;
+                                if total + reclen as u32 > size {
+                                    break;
+                                }
+                                total += reclen as u32;
+                                unsafe {
+                                    (*(dirp_ptr as *mut dirent64)).d_ino = 1;
+                                    (*(dirp_ptr as *mut dirent64)).d_off = 2;
+                                    (*(dirp_ptr as *mut dirent64)).d_reclen = reclen;
+                                    (*(dirp_ptr as *mut dirent64)).d_type =
+                                        match value.1.as_bytes()[0] {
```

Comment:
I think we don't need to use unsafe here.

Reply:
How to do